### PR TITLE
feat(dup_enhancement#17):  replica follower load duplication data when open replica

### PR DESCRIPTION
**Pull request by Shuo-Jia**
# Ref-Issue
https://github.com/apache/incubator-pegasus/issues/892

# Change

follower load duplication data if it's a replica follower

### CODE REVIEW

File: src/server/pegasus_server_impl.cpp [hunk 1]
```diff
@@ -1496,62 +1498,67 @@ ::dsn::error_code pegasus_server_impl::start(int argc, char **argv)
     //          3, restore_dir is exist
     //
     bool db_exist = true;
-    auto path = ::dsn::utils::filesystem::path_combine(data_dir(), "rdb");
-    if (::dsn::utils::filesystem::path_exists(path)) {
+    auto rdb_path = ::dsn::utils::filesystem::path_combine(data_dir(), "rdb");
```
Comment (hycdong) [inline src/server/pegasus_server_impl.cpp:1501]: L1500 - L1559 have too many condiftions, doing following checks:
- db already exist
- db not exist (1. normal 2. duplication 3. restore)

Could you please refactor it?

Comment (Shuo-Jia) [reply src/server/pegasus_server_impl.cpp:1501]: `refactor` is not this pr, which may introduce some other bug

Comment (hycdong) [reply src/server/pegasus_server_impl.cpp:1501]: okay, expecting your further refactor pull request

**Review by hycdong — APPROVED** [review]

File: src/server/pegasus_server_impl.cpp [hunk 2]
```diff
@@ -1496,62 +1498,67 @@ ::dsn::error_code pegasus_server_impl::start(int argc, char **argv)
...
+                ddebug_replica("try to restore from restore_dir = {}", restore_dir);
+                if (::dsn::utils::filesystem::directory_exists(restore_dir)) {
+                    // here, we just rename restore_dir to rdb, then continue the normal process
+                    if (::dsn::utils::filesystem::rename_path(restore_dir, rdb_path)) {
+                        ddebug_replica(
+                            "rename restore_dir({}) to rdb({}) succeed", restore_dir, rdb_path);
+                    } else {
+                        derror_replica(
+                            "rename restore_dir({}) to rdb({}) failed", restore_dir, rdb_path);
+                        return ::dsn::ERR_FILE_OPERATION_FAILED;
```
Comment (levy5307) [inline src/server/pegasus_server_impl.cpp:1538]: ::dsn::ERR_FILE_OPERATION_FAILED ==> dsn::ERR_FILE_OPERATION_FAILED

File: src/server/pegasus_server_impl.cpp [hunk 3]
```diff
@@ -1496,62 +1498,67 @@ ::dsn::error_code pegasus_server_impl::start(int argc, char **argv)
...
-            // case 2
-            if (force_restore) {
-                derror("%s: try to restore, but we can't combine restore_dir from envs",
-                       replica_name());
-                return ::dsn::ERR_FILE_OPERATION_FAILED;
-            } else {
-                db_exist = false;
-                dinfo("%s: open a new db, path = %s", replica_name(), path.c_str());
+        // case 2
+        if (dsn::utils::filesystem::path_exists(duplication_path) && is_duplication_follower()) {
```
Comment (levy5307) [inline src/server/pegasus_server_impl.cpp:1508]: Maybe you can split these code into several small functions. Each function deal with one case.
You can do it in later pull requests.

File: src/server/pegasus_server_impl.cpp [hunk 4]
```diff
@@ -1496,62 +1498,67 @@ ::dsn::error_code pegasus_server_impl::start(int argc, char **argv)
     //          3, restore_dir is exist
     //
     bool db_exist = true;
-    auto path = ::dsn::utils::filesystem::path_combine(data_dir(), "rdb");
-    if (::dsn::utils::filesystem::path_exists(path)) {
+    auto rdb_path = ::dsn::utils::filesystem::path_combine(data_dir(), "rdb");
+    auto duplication_path = duplication_dir();
+    if (::dsn::utils::filesystem::path_exists(rdb_path)) {
         // only case 1
```
Comment (levy5307) [inline src/server/pegasus_server_impl.cpp:1504]: Add detail description to show what does case 1 mean:
```
// case 1: we open the db that already exist
```

File: src/server/pegasus_server_impl.cpp (hunk 3 above)
Comment (Shuo-Jia) [reply src/server/pegasus_server_impl.cpp:1508]: Yeah, it's good

File: src/server/pegasus_server_impl.cpp (hunk 4 above)
Comment (Shuo-Jia) [reply src/server/pegasus_server_impl.cpp:1504]: Has comment in code start: https://github.com/apache/incubator-pegasus/pull/917/files/c07465e1ddb1831b0953bce1f71c64b234a5c686#diff-951c6edb3ffd9c5cea7460f0cbdb7ebc1bd792615bf4d8415379665a872db06eR1489

**Review by levy5307 — APPROVED** [review]

**Pull request merged**